topdown/cache/cache.go

```diff
@@ -68,23 +99,44 @@ type InterQueryCacheValue interface {
 type InterQueryCache interface {
 	Get(key ast.Value) (value InterQueryCacheValue, found bool)
 	Insert(key ast.Value, value InterQueryCacheValue) int
+	InsertWithExpiry(key ast.Value, value InterQueryCacheValue, expiresAt time.Time) int
 	Delete(key ast.Value)
 	UpdateConfig(config *Config)
 	Clone(value InterQueryCacheValue) (InterQueryCacheValue, error)
 }
 
 // NewInterQueryCache returns a new inter-query cache.
 func NewInterQueryCache(config *Config) InterQueryCache {
-	return &cache{
-		items:  map[string]cacheItem{},
-		usage:  0,
-		config: config,
-		l:      list.New(),
+	return newCache(config)
+}
+
+// NewInterQueryCacheWithContext returns a new inter-query cache with context.
+func NewInterQueryCacheWithContext(ctx context.Context, config *Config) InterQueryCache {
+	iqCache := newCache(config)
+
+	// Start routine to clean up stale values once every StaleEntryEvictionPeriodSeconds
+	cleanupPeriod := iqCache.staleEntryEvictionTimePeriodSeconds()
+	if cleanupPeriod > 0 {
+		ticker := time.NewTicker(time.Duration(cleanupPeriod) * time.Second)
+		go func() {
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
```

Comment:
What would happen if we still in the middle of a clean up and ticker ticks? We should probably stop the ticker, do the work and recreate it.

Reply:
Currently we have `c.mtx.Lock()` ensuring only one clean up routine runs at a time. But the time period `StaleEntryEvictionPeriodSeconds` is enforced between start of each routine. We should reset the ticker if we intend to enforce it between end of previous routine and start of new one.

Reply:
@ashutosh-narkar I have added a recreate ticker step in the cleanup routine's defer procedure. Don't think an explicit `Stop()` call is required. Please do let me know if it LGTY.